design-proposals/primary_network_binding_for_seamless_migration.md

```diff
@@ -0,0 +1,596 @@
+work on a primary pod network. The masquerade, bridge and passt bindings
+work with the primary binding.
+The SR-IOV and macvtap bindings are examples of the ones which cannot be
+used on the primary pod network.
+
+> **Note**: Consuming a CNI IPAM requirement for the primary pod network
+> is not a hard requirement. There are exceptions to this, e.g. instead
+> of consuming the CNI IPAM, the guest can consume an IP from the
+> network-provider, set statically through cloud-init or other means.
+
+### Migration Support
+
+Kubevirt allows a VM to migrate only if it uses the masquerade binding
+or a plugin binding that is marked
+as [supporting migration](https://kubevirt.io/user-guide/network/network_binding_plugins/#migration)
+for its primary pod network (or if the primary is not used at all).
+The ability to migrate is agnostic to the support for seamless migration
+with no traffic degradation. In fact, the primary pod network traffic is
+currently expected to get broken during the migration (e.g. with
+masquerade, the pod IP change/officials).
+
+Users are recommended to use the secondary network with a L2
+connectivity to preserve network connectivity with minimal impact. This
+also comes with a limitation of not attempting to manage IP on secondary
+networks, due to dependencies on sticky IP allocations for the pods up
+to the ability to pass the IP allocation down to the VM itself.
+
+The proposed solutions need to take into account the challenges the
+existing known bindings have with the primary pod network and with migration.
+
+## Proposal
+
+The following proposal attempts to provide a solution from VMs that use
+the primary pod network for the majority of the traffic and require a
+seamless migration with no (or minimal) traffic connectivity
+interruption.
+
+The solution should consider two important points:
+
+- Be agnostic to a specific network provided and CNI, by defining the
+  requirements well.
+- Co-exit with the current solutions without adding too much burden on
+  maintenance. E.g. the logic and codebase should not diverge too much
+  from the existing core network flow.
+
+### Pre Requirements
+
+The main challenge when working with the primary pod network is to
+assure connectivity recovers after a VM migration. A new requirement is
+added with this proposal, to preserve existing connectivity (e.g. TCP
+connection) with minimal (to none) disruption.
+
+As we are focusing here on the primary pod network, these requirements
+translate to several pre-requirements from the network provider & CNI:
+
+- Sticky IP: An IPAM plugin assigns an IP for each pod. In order to
+  support migration and preserve seamless connectivity, the pod IP on
+  the target node must use the same IP.
+  The network-provider/CNI/IPAM needs to assure such an option exists so
+  the same IP used for the primary pod network can be moved to the
+  target pod.
+- Flat Cluster Subnet: In order to allow the same IP to be assigned on a
```

Comment:
> All I am asking here is to clarify what subnet you are talking about.

Done.

> Additionally, do we really need flat subnet? I could imagine a working scenario (hacky for sure) where you could persist the IP across nodes without flat subnet. So this point is just specification of the first one and my question would be if that is really needed.

If we just preserve the IP but not the subnet, then the pod on the other node will use an IP that is not included in that node subnet, therefore it will not be able to communicate.
In this context, a subnet is equivalent to a LAN.